File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-c/ext_break_find_first__submitted.c
/*
 * ext_break_find_first - find-first break pattern (TSVC s481)
 *
 * The reference NumPy implementation (Python) is:
 *   for i in range(LEN_1D):
 *       if d[i] < 0.0:
 *           break
 *       a[i] = a[i] + b[i] * c[i]
 *
 * The benchmark harness passes four 1-D contiguous arrays of the same length and an
 * integer length. All arrays are distinct (the reference creates them separately) and are
 * double-precision (float64). The kernel must update a in-place and perform the arithmetic only
 * up to (but not including) the first negative entry in d.
 *
 * Strategy
 * --------
 * * First locate the break index with a simple scalar scan. This scan is cheap - it executes at
 *   most LEN_1D iterations and exits as soon as the negative element is found.
 * * Once the index is known we can run the remaining arithmetic as a separate loop that the
 *   compiler can fully vectorise and, optionally, parallelise with OpenMP. Because the break
 *   index is known beforehand there is no data-dependent branch inside the compute loop, allowing
 *   the optimiser to emit SIMD instructions.
 * * The implementation uses restrict qualifiers (the benchmark guarantees that the four
 *   arrays do not alias) and a parallel for with static scheduling - a cheap way to get
 *   multi-core speed-up when the break occurs far into the array. If the break occurs very early
 *   the overhead of the OpenMP team creation is negligible compared with the baseline serial
 *   version.
 *
 * The function name and signature follow the benchmark ABI: the kernel name is exactly the
 * task key (the harness looks for a function with this name) and the arguments are pointers to the
 * data followed by the length. All pointers are to double and the length is an int64_t.
 * The symbol name must be ext_break_find_first_fp64 and include workspace arguments (ignored here).
 */

#include <omp.h>
#include <stdint.h>

/*
 * Kernel entry point.
 *   a  - input/output array (modified in place)
 *   b  - input array
 *   c  - input array
 *   d  - input array containing a single negative element defining the break point
 *   LEN_1D - number of elements in each array
 *   workspace - ignored pointer
 *   workspace_bytes - ignored size
 */
void ext_break_find_first_fp64(double * restrict a,
                         const double * restrict b,
                         const double * restrict c,
                         const double * restrict d,
                         int64_t LEN_1D,
                         uint8_t * restrict workspace,
                         int64_t workspace_bytes)
{
    // Avoid unused parameter warnings
    (void)workspace;
    (void)workspace_bytes;

    int64_t break_idx = LEN_1D; // default: no break found (should not happen in the benchmark)

    // Find the first negative value in d
    for (int64_t i = 0; i < LEN_1D; ++i) {
        if (d[i] < 0.0) {
            break_idx = i;
            break;
        }
    }

    // Compute the expression up to break_idx. The loop can be vectorised and, for large
    // break_idx, parallelised across threads.
    #pragma omp parallel for schedule(static) if(break_idx > 0)
    for (int64_t i = 0; i < break_idx; ++i) {
        a[i] = a[i] + b[i] * c[i];
    }
}

File: paper_artifacts/experiments/llr9/artifacts/sources/oss120b-fortran/ext_break_find_first__submitted.f90
! *********************************************************************
! ext_break_find_first kernel for HPCAgent-Bench (Fortran version)
! ---------------------------------------------------------------------
! The kernel receives four input arrays `a`, `b`, `c`, `d` each of length LEN_1D.
! It scans `d` for the first negative element, breaks the loop when such
! an element is found, and for all preceding elements updates `a` with
! a[i] = a[i] + b[i] * c[i].
!
! The reference implementation (Python/NumPy) performs a serial scan with an
! early break. For large arrays, a parallel two‑pass algorithm can achieve a
! speed‑up while preserving the same semantics.
!
! The subroutine follows the name and argument order expected by the
! benchmark harness:
!   subroutine ext_break_find_first(a, b, c, d, LEN_1D)
! where
!   a, b, c, d : real(8) arrays, size LEN_1D,
!                intent(inout) for a, intent(in) for the others,
!   LEN_1D    : integer(8) scalar, intent(in).
!
! The implementation uses OpenMP to find the first negative index in `d`
! via a parallel reduction and then updates the prefix of `a` in a parallel
! loop. This approach scales with the number of cores while preserving the
! benchmark's required semantics.
! *********************************************************************

subroutine ext_break_find_first_fp64(a, b, c, d, LEN_1D, workspace, workspace_bytes) bind(C, name="ext_break_find_first_fp64")
    use iso_c_binding, only: c_int64_t, c_double, c_ptr
    use omp_lib
    implicit none

    ! Arguments
    integer(c_int64_t), value :: LEN_1D
    type(c_ptr), value :: workspace
    integer(c_int64_t), value :: workspace_bytes
    real(c_double), intent(inout) :: a(*)
    real(c_double), intent(in) :: b(*)
    real(c_double), intent(in) :: c(*)
    real(c_double), intent(in) :: d(*)

    ! Local variables
    integer(c_int64_t) :: i
    integer(c_int64_t) :: break_idx
    integer(c_int64_t), parameter :: sentinel = huge(0_c_int64_t)

    ! Edge case: empty array – nothing to do
    if (LEN_1D <= 0_c_int64_t) then
        return
    end if

    ! -----------------------------------------------------------------
    ! Phase 1: Find the smallest index i where d(i) < 0 using a parallel
    ! reduction with the MIN operator. The reduction's private copies are
    ! initialised to the maximum representable integer, so after the region
    ! `break_idx` holds the minimum matching index or a sentinel value if no
    ! element satisfies the condition.
    ! -----------------------------------------------------------------
    break_idx = sentinel
    !$omp parallel default(none) private(i) reduction(min:break_idx) shared(d, LEN_1D)
    !$omp do schedule(static)
    do i = 1, LEN_1D
        if (d(i) < 0.0_c_double) then
            break_idx = i
        end if
    end do
    !$omp end do
    !$omp end parallel

    ! If no negative was found (should not happen for the benchmark's
    ! generated inputs except for LEN_1D <= 1), treat it as a break after the
    ! last element.
    if (break_idx > LEN_1D) then
        break_idx = LEN_1D + 1_c_int64_t
    end if

    ! -----------------------------------------------------------------
    ! Phase 2: Update a(i) for all indices before the break point.
    ! -----------------------------------------------------------------
    if (break_idx > 1_c_int64_t) then
        !$omp parallel do default(none) private(i) shared(a, b, c, break_idx)
        do i = 1, break_idx - 1
            a(i) = a(i) + b(i) * c(i)
        end do
        !$omp end parallel do
    end if

    ! No explicit return value – results are stored in the updated array `a`.
end subroutine ext_break_find_first_fp64
